### PR TITLE
feature(events): add behavior-state-changed event

### PR DESCRIPTION
**Pull request by okke-formsma**
This commit adds a behavior-state-changed event which is called just before a behavior is activated. The last listener (in `behavior.c`) actually calls the behavior.

This set of commits leads up to a cleanup of the entire behavior handling, making complex behaviors like hold-tap and sticky keys a little simpler. The full set of planned commits can be found at https://github.com/zmkfirmware/zmk/pull/532.

TLDR;
We currently have the following events

* position_state_changed
    * keycode_state_changed
    * layer_state_changed

What is missing from this event flow are behavior bindings. This means they can not be observed like other events.

* position_state_changed
  * behavior_state_changed
    * keycode_state_changed
    * layer_state_changed

### CODE REVIEW

**Comment by okke-formsma** [comment]
Why is this change needed? It's needed so we can create behaviors based on behavior events, such as combos that trigger on <&kp A, &kp B> instead of positions <1, 2>.

**Comment by okke-formsma** [comment]
Rebased.

**Comment by okke-formsma** [comment]
@petejohanson could you have a look at this?

File: app/include/zmk/events/behavior_state_changed.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * SPDX-License-Identifier: MIT
+ */
+
+#pragma once
+
+#include <zephyr.h>
+#include <zmk/event_manager.h>
+#include <zmk/behavior.h>
+
+struct zmk_behavior_state_changed {
```
Comment (petejohanson) [inline app/include/zmk/events/behavior_state_changed.h:13]: This struct and the existing binding event struct are duplicative and confusing after this addition. It seems like we should be unifying this as part of this work.

File: app/src/behavior.c [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+#include <zmk/event_manager.h>
+#include <zmk/events/position_state_changed.h>
+#include <zmk/events/behavior_state_changed.h>
+#include <zmk/events/sensor_event.h>
+
+int zmk_behavior_state_changed_listener(const zmk_event_t *zmk_ev) {
+    struct zmk_behavior_state_changed *ev = as_zmk_behavior_state_changed(zmk_ev);
+    if (ev == NULL) {
+        return 0;
```
Comment (petejohanson) [inline app/src/behavior.c:25]: If we're getting any other event type, something is really amiss, and I feel like we should be returning an error (we might not be doing this elsewhere, but we should), I would suggest `-ENOTSUP` or somesuch.

File: app/src/behavior.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    };
+
+    int err = behavior_keymap_binding_convert_central_state_dependent_params(&ev->binding, event);
+    if (err) {
+        LOG_ERR("Failed to convert relative to absolute behavior binding (err %d)", err);
+        return err;
+    }
+
+    int ret;
+    if (ev->pressed) {
```
Comment (petejohanson) [inline app/src/behavior.c:40]: Perhaps the behavior driver API needs to evolve too.... I'm on the fence about explicit pressed/released functions, versus generic "state changed" function, which gives us more flexibility down the road.

**Review by petejohanson — COMMENTED** [review]
A few thoughts, thanks for the ping.

File: app/src/behavior.c (hunk 3 above)
Comment (okke-formsma) [reply app/src/behavior.c:40]: In https://github.com/zmkfirmware/zmk/issues/794 I argue that we should have explicit pressed/released event handler functions for all events for which the press/release duality exists, as most users of events dispatch to a different handler function based on the `state` now anyway.

File: app/include/zmk/events/behavior_state_changed.h (hunk 1 above)
Comment (okke-formsma) [reply app/include/zmk/events/behavior_state_changed.h:13]: I feel `zmk_behavior_binding` is a valid type on it's own. It's a nice succinct description of the things that exist in a keymap:

```
struct zmk_behavior_binding {
    char *behavior_dev;
    uint32_t param1;
    uint32_t param2;
};
```

Comment (okke-formsma) [reply app/include/zmk/events/behavior_state_changed.h:13]: On second thought, were you talking about this?

```
struct zmk_behavior_binding_event {
    int layer;
    uint32_t position;
    int64_t timestamp;
};
```

Yeah, that one and `zmk_behavior_state_changed` should definitely be unified.

Comment (petejohanson) [reply app/include/zmk/events/behavior_state_changed.h:13]: Yes exactly.

File: app/src/behavior.c (hunk 2 above)
Comment (petejohanson) [reply app/src/behavior.c:25]: Bump.

**Review by petejohanson — COMMENTED** [review]
@okke-formsma Any idea on if/when you might be able to get back to this

This and #547 will certainly conflict, so considering ordering/timing.
